File: examples/complete-demo/utils/blockchain.py
"""
Blockchain utilities for ERC-8004 demo
"""
import os
from web3 import Web3
from eth_account import Account
import json

class BlockchainClient:
    def __init__(self, rpc_url=None, private_key=None):
        self.rpc_url = rpc_url or os.getenv('RPC_URL', 'http://localhost:8545')
        self.private_key = private_key or os.getenv('PRIVATE_KEY')
        
        # Initialize Web3
        self.w3 = Web3(Web3.HTTPProvider(self.rpc_url))
        
        if not self.w3.is_connected():
            raise Exception(f"Failed to connect to blockchain at {self.rpc_url}")
        
        # Setup account
        if self.private_key:
            self.account = Account.from_key(self.private_key)
            self.address = self.account.address
        else:
            raise Exception("Private key is required")
    
    def get_contract(self, address, abi):
        """Get contract instance"""
        return self.w3.eth.contract(address=address, abi=abi)
    
    def send_transaction(self, transaction):
        """Send a transaction"""
        # Estimate gas
        transaction['gas'] = self.w3.eth.estimate_gas(transaction)
        
        # Get nonce
        transaction['nonce'] = self.w3.eth.get_transaction_count(self.address)
        
        # Sign transaction
        signed_txn = self.w3.eth.account.sign_transaction(transaction, self.private_key)
        
        # Send transaction
        tx_hash = self.w3.eth.send_raw_transaction(signed_txn.raw_transaction)
        
        # Wait for receipt
        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
        return receipt
    
    def load_contract_abi(self, contract_name):
        """Load contract ABI from compiled artifacts"""
        abi_path = f"contracts/out/{contract_name}.sol/{contract_name}.json"
        
        if os.path.exists(abi_path):
            with open(abi_path, 'r') as f:
                artifact = json.load(f)
                return artifact['abi']
        else:
            # Fallback to minimal ABI for demo
            return self._get_minimal_abi(contract_name)
    
    def _get_minimal_abi(self, contract_name):
        """Minimal ABIs for demo purposes"""
        abis = {
            'IdentityRegistry': [
                {
                    "inputs": [{"type": "string", "name": "tokenURI_"}],
                    "name": "register",
                    "outputs": [{"type": "uint256", "name": ""}],
                    "stateMutability": "nonpayable",
                    "type": "function"
                },
                {
                    "inputs": [{"type": "uint256", "name": "agentId"}],
                    "name": "tokenURI", 
                    "outputs": [{"type": "string", "name": ""}],
                    "stateMutability": "view",
                    "type": "function"
                }
            ],
            'ReputationRegistry': [
                {
                    "inputs": [
                        {"type": "uint256", "name": "agentId"},
                        {"type": "uint8", "name": "score"},
                        {"type": "string[]", "name": "tags"},
                        {"type": "string", "name": "reportURI"},
                        {"type": "bytes32", "name": "reportHash"},
                        {"type": "bytes", "name": "authorization"}
                    ],
                    "name": "submitFeedback",
                    "outputs": [],
                    "stateMutability": "nonpayable",
                    "type": "function"
                },
                {
                    "inputs": [{"type": "uint256", "name": "agentId"}],
                    "name": "getAverageScore",
                    "outputs": [{"type": "uint256", "name": ""}],
                    "stateMutability": "view",
                    "type": "function"
                }
            ],
            'ValidationRegistry': [
                {
                    "inputs": [
                        {"type": "bytes32", "name": "requestHash"},
                        {"type": "string", "name": "validationType"}, 
                        {"type": "string", "name": "requestURI"}
                    ],
                    "name": "requestValidation",
                    "outputs": [],
                    "stateMutability": "nonpayable", 
                    "type": "function"
                },
                {
                    "inputs": [
                        {"type": "bytes32", "name": "requestHash"},
                        {"type": "uint8", "name": "result"},
                        {"type": "string", "name": "responseURI"},
                        {"type": "bytes32", "name": "responseHash"}
                    ],
                    "name": "submitValidation",
                    "outputs": [],
                    "stateMutability": "nonpayable",
                    "type": "function"
                }
            ]
        }
        
        return abis.get(contract_name, [])

File: examples/complete-demo/src/utils/blockchain.ts
import { ethers } from 'ethers';
import { ContractABI, NetworkConfig, TransactionReceipt } from '../types';

export class BlockchainClient {
  private provider: ethers.JsonRpcProvider;
  private wallet: ethers.Wallet;
  
  constructor(
    private config: NetworkConfig,
    privateKey: string
  ) {
    this.provider = new ethers.JsonRpcProvider(config.rpcUrl);
    this.wallet = new ethers.Wallet(privateKey, this.provider);
  }

  async isConnected(): Promise<boolean> {
    try {
      await this.provider.getNetwork();
      return true;
    } catch (error) {
      return false;
    }
  }

  getAddress(): string {
    return this.wallet.address;
  }

  async getBalance(): Promise<string> {
    const balance = await this.provider.getBalance(this.wallet.address);
    return ethers.formatEther(balance);
  }

  getContract(address: string, abi: ContractABI[]): ethers.Contract {
    return new ethers.Contract(address, abi, this.wallet);
  }

  async sendTransaction(
    contract: ethers.Contract,
    functionName: string,
    args: any[] = [],
    overrides: any = {}
  ): Promise<TransactionReceipt> {
    try {
      const tx = await contract[functionName](...args, {
        gasLimit: 500000,
        ...overrides
      });
      
      const receipt = await tx.wait();
      
      return {
        transactionHash: receipt.hash,
        blockNumber: receipt.blockNumber,
        gasUsed: receipt.gasUsed.toString(),
        status: receipt.status
      };
    } catch (error) {
      throw new Error(`Transaction failed: ${error}`);
    }
  }

  async callFunction(
    contract: ethers.Contract,
    functionName: string,
    args: any[] = []
  ): Promise<any> {
    try {
      return await contract[functionName](...args);
    } catch (error) {
      throw new Error(`Contract call failed: ${error}`);
    }
  }

  // Contract ABIs for the ERC-8004 registries
  static getContractABI(contractName: string): ContractABI[] {
    const abis: Record<string, ContractABI[]> = {
      IdentityRegistry: [
        {
          inputs: [{ internalType: 'string', name: 'tokenURI_', type: 'string' }],
          name: 'register',
          outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
          stateMutability: 'nonpayable',
          type: 'function'
        },
        {
          inputs: [{ internalType: 'uint256', name: 'agentId', type: 'uint256' }],
          name: 'tokenURI',
          outputs: [{ internalType: 'string', name: '', type: 'string' }],
          stateMutability: 'view',
          type: 'function'
        },
        {
          inputs: [{ internalType: 'uint256', name: 'agentId', type: 'uint256' }],
          name: 'ownerOf',
          outputs: [{ internalType: 'address', name: '', type: 'address' }],
          stateMutability: 'view',
          type: 'function'
        },
        {
          inputs: [{ internalType: 'uint256', name: 'agentId', type: 'uint256' }],
          name: 'agentExists',
          outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
          stateMutability: 'view',
          type: 'function'
        }
      ],
      ReputationRegistry: [
        {
          inputs: [
            { internalType: 'uint256', name: 'agentId', type: 'uint256' },
            { internalType: 'uint8', name: 'score', type: 'uint8' },
            { internalType: 'string[]', name: 'tags', type: 'string[]' },
            { internalType: 'string', name: 'reportURI', type: 'string' },
            { internalType: 'bytes32', name: 'reportHash', type: 'bytes32' },
            { internalType: 'bytes', name: 'authorization', type: 'bytes' }
          ],
          name: 'submitFeedback',
          outputs: [],
          stateMutability: 'nonpayable',
          type: 'function'
        },
        {
          inputs: [{ internalType: 'uint256', name: 'agentId', type: 'uint256' }],
          name: 'getAverageScore',
          outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
          stateMutability: 'view',
          type: 'function'
        },
        {
          inputs: [{ internalType: 'uint256', name: 'agentId', type: 'uint256' }],
          name: 'getFeedbackCount',
          outputs: [{ internalType: 'uint256', name: '', type: 'uint256' }],
          stateMutability: 'view',
          type: 'function'
        }
      ],
      ValidationRegistry: [
        {
          inputs: [
            { internalType: 'bytes32', name: 'requestHash', type: 'bytes32' },
            { internalType: 'string', name: 'validationType', type: 'string' },
            { internalType: 'string', name: 'requestURI', type: 'string' }
          ],
          name: 'requestValidation',
          outputs: [],
          stateMutability: 'nonpayable',
          type: 'function'
        },
        {
          inputs: [
            { internalType: 'bytes32', name: 'requestHash', type: 'bytes32' },
            { internalType: 'uint8', name: 'result', type: 'uint8' },
            { internalType: 'string', name: 'responseURI', type: 'string' },
            { internalType: 'bytes32', name: 'responseHash', type: 'bytes32' }
          ],
          name: 'submitValidation',
          outputs: [],
          stateMutability: 'nonpayable',
          type: 'function'
        },
        {
          inputs: [{ internalType: 'bytes32', name: 'requestHash', type: 'bytes32' }],
          name: 'getConsensusResult',
          outputs: [{ internalType: 'uint8', name: '', type: 'uint8' }],
          stateMutability: 'view',
          type: 'function'
        }
      ]
    };

    return abis[contractName] || [];
  }

  // Utility functions
  static hashString(input: string): string {
    return ethers.keccak256(ethers.toUtf8Bytes(input));
  }

  static hashObject(obj: any): string {
    const jsonString = JSON.stringify(obj, Object.keys(obj).sort());
    return ethers.keccak256(ethers.toUtf8Bytes(jsonString));
  }

  async signMessage(message: string): Promise<string> {
    return await this.wallet.signMessage(message);
  }

  static verifySignature(message: string, signature: string, expectedAddress: string): boolean {
    try {
      const recoveredAddress = ethers.verifyMessage(message, signature);
      return recoveredAddress.toLowerCase() === expectedAddress.toLowerCase();
    } catch {
      return false;
    }
  }
}